### PR TITLE
[StructuralMechanicsAplication] Clean up prism neighbours + test

### PR DESCRIPTION
**Pull request by loumalouomega**

### CODE REVIEW

**Comment by loumalouomega** [comment]
One think until now the number of cpp_test in @KratosMultiphysics/structural-mechanics was 0. I think is a good idea to implement more cpp test, these tests just check the work of one specific algorithm or utility

File: applications/MeshingApplication/CMakeLists.txt [hunk 1]
```diff
@@ -64,7 +64,7 @@ endif(${KRATOS_BUILD_TESTING} MATCHES ON)
 ###############################################################
 ## define library Kratos which defines the basic python interface
 add_library(KratosMeshingApplication SHARED ${KRATOS_MESHING_APPLICATION_SOURCES} ${KRATOS_MESHING_APPLICATION_TESTING_SOURCES})
-target_link_libraries(KratosMeshingApplication KratosCore triangle tetMeshOpt)
+target_link_libraries(KratosMeshingApplication KratosCore KratosStructuralMechanicsApplication triangle tetMeshOpt)
```
Comment (philbucher) [inline applications/MeshingApplication/CMakeLists.txt]: is this a good idea?
Does one then have to always compile StructuralMechanics to compile the MeshingApplication?

**Review by philbucher — COMMENTED** [review]
I am not familiar with your implementation, looks ok to me.
maybe @RiccardoRossi can have a look?

And thanks a lot for adding the cpp tests!

**Comment by philbucher** [comment]
@roigcarlo are the cpp tests in the applications also executed by travis or only the small python tests?

File: applications/MeshingApplication/CMakeLists.txt (hunk 1 above)
Comment (loumalouomega) [reply applications/MeshingApplication/CMakeLists.txt]: Certainly I don't know, that is a question I have to @roigcarlo, in any case the other alternative would be to remove the utility that uses the process in the meshing application (it is deprecated)

Comment (loumalouomega) [reply applications/MeshingApplication/CMakeLists.txt]: Done, now it compiles the utility just when the StructuralMechanics is compiled

**Comment by loumalouomega** [comment]
@philbucher Travis runs both tests

File: applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 2]
```diff
@@ -25,7 +25,9 @@
 
 #include "custom_utilities/local_refine_triangle_mesh.hpp"
 #include "custom_utilities/local_refine_prism_mesh.hpp"
+#ifdef  STRUCTURAL_MECHANICS_APPLICATION
```
Comment (RiccardoRossi) [inline applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp]: why is this needed?

Comment (RiccardoRossi) [reply applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp]: note that you should REALLY AVOID cross including this file

File: applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h [hunk 3]
```diff
@@ -0,0 +1,306 @@
+// KRATOS  ___|  |                   |                   |
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h:1]: is this file specific to the StructuralMechanicsApp?

File: kratos/includes/key_hash.h [hunk 4]
```diff
@@ -50,22 +56,21 @@ namespace Kratos
...
-     * @param TClassType The type of class to be hashed
+     * @brief This method combines hash until it reaches the last class in order to obtain a corresponding seed
+     * @tparam TClassType The type of class to be hashed
      * @param First The first class to be compared
      * @param Last The last class to be compared
      * @return The resulting seed
      */
     template <class TClassType>
-    inline std::size_t HashRange(
+    inline HasType HashRange(
```
Comment (RiccardoRossi) [inline kratos/includes/key_hash.h]: I think it should be HashType not HasType

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
i think there is something wrong with cross including between different apps. Also you miss a test

File: applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+///@name Kratos Classes
+///@{
+
+/**
+ * @class PrismNeighboursProcess
+ * @ingroup StructuralMechanicsApplication
+ * @brief An algorithm that looks for neighbour nodes and elements in a mesh of prismatic elements
+ * @details For that pourpose if builds an unordered map of the surrounding elements and nodes and performs different checks
+ * @author Vicente Mataix Ferrandiz
+ * @todo Remove the dependence of the boost::unordered_map
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h]: I think this todo can be removed

**Comment by loumalouomega** [comment]
@RiccardoRossi  I added a test?, what do you mean?

File: applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h (hunk 5 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h]: Yes

File: kratos/includes/key_hash.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/includes/key_hash.h]: True

File: applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h (hunk 3 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h:1]: yes, do you want me to move it to the core?

File: applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp]: I can move it to the core or reverting to only header file

Comment (loumalouomega) [reply applications/MeshingApplication/custom_python/add_custom_utilities_to_python.cpp]: Or I can remove the utility, which with MMG is deprecated

File: applications/MeshingApplication/CMakeLists.txt [hunk 6]
```diff
@@ -64,7 +64,11 @@ endif(${KRATOS_BUILD_TESTING} MATCHES ON)
 ###############################################################
 ## define library Kratos which defines the basic python interface
 add_library(KratosMeshingApplication SHARED ${KRATOS_MESHING_APPLICATION_SOURCES} ${KRATOS_MESHING_APPLICATION_TESTING_SOURCES})
-target_link_libraries(KratosMeshingApplication KratosCore triangle tetMeshOpt)
+if(${STRUCTURAL_MECHANICS_APPLICATION} MATCHES ON)
```
Comment (pooyan-dadvand) [inline applications/MeshingApplication/CMakeLists.txt]: why you are adding dependency between two apps?

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Please remove the dependency between meshing application and structural

File: applications/MeshingApplication/CMakeLists.txt (hunk 6 above)
Comment (loumalouomega) [reply applications/MeshingApplication/CMakeLists.txt]: The dependency was already there, but this didn't happen when the process was header only. As I said the utility using the process is deprecated, so I can remove it instead.

**Comment by loumalouomega** [comment]
Done

File: applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h (hunk 3 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_processes/prism_neighbours_process.h:1]: i reformulate the question:

"what of this is specific of the StructuralMechanics? can it be useful to other apps"?

**Review by RiccardoRossi — COMMENTED** [review]
@loumalouomega i am willing to approve, however please confirm me that there is now no interdependency between the applications

**Comment by loumalouomega** [comment]
The dependency is removed, can be used in other apps in the same way that the utilities/processes to search the neighbours of the tetrahedra an the triangles, but for prisms

**Review by RiccardoRossi — APPROVED** [review]

**Review by pooyan-dadvand — APPROVED** [review]
